### PR TITLE
Andrey Rohau / Create test, GetAll + validation, Updated db + migration

### PR DESCRIPTION
**Pull request by andreyrogov8**

### CODE REVIEW

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 1]
```diff
@@ -24,25 +29,227 @@ public class Handler : IRequestHandler<Request, Response>
...
+            public Handler(IApplicationDbContext context, IMapper mapper, UserManager<ApplicationUser> userManager)
             {
                 _context = context;
                 _mapper = mapper;
+                _userManager = userManager;
+
             }
             public async Task<Response> Handle(Request request, CancellationToken cancelationtoken)
             {
+                if (await _userManager.Users.FirstOrDefaultAsync(x => x.Id == request.UserId).ConfigureAwait(false) == null)
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:46]: _userManager.Users.Any(x => x.Id == request.UserId)

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 2]
```diff
@@ -24,25 +29,227 @@ public class Handler : IRequestHandler<Request, Response>
...
+                {
+                    return new Response()
+                    {
+                        Id = null,
+                        Message = "UserId is not found"
+                    };
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:52]: In this case instead response I supposed using exception e.g. throw EntityNotFoundException("UserId is not found"). Or you can create your own exception e.g. ValidationException or BusinessException

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 3]
```diff
@@ -24,25 +29,227 @@ public class Handler : IRequestHandler<Request, Response>
...
+                }
+
+                var Hr = await _userManager.Users.FirstOrDefaultAsync(x => x.Id == request.HrId).ConfigureAwait(false);
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:55]: Any()

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 4]
```diff
@@ -24,25 +29,227 @@ public class Handler : IRequestHandler<Request, Response>
...
+
+                if (request.HrId.HasValue
+                    && Hr == null)
+                {
+                    return new Response()
+                    {
+                        Id = null,
+                        Message = "HrId is not found"
+                    };
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:64]: throw appropriate exception

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 5]
```diff
@@ -24,25 +29,227 @@ public class Handler : IRequestHandler<Request, Response>
...
+                }
+                if (!(await _userManager.GetRolesAsync(Hr)).Contains("HumanResourceManager"))
+                {
+                    return new Response()
+                    {
+                        Id = null,
+                        isHr = true
+                    };
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:72]: throw appropriate exception

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 6]
```diff
@@ -24,25 +29,227 @@ public class Handler : IRequestHandler<Request, Response>
...
+                }
+                if(request.HrId == request.UserId)
+                {
+                    return new Response()
+                    {
+                        Id = null,
+                        Message = "You can't assign yourself"
+                    };
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:80]: throw appropriate exception

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 7]
```diff
@@ -24,25 +29,227 @@ public class Handler : IRequestHandler<Request, Response>
...
+                var v = await _context.Tests.FirstOrDefaultAsync(x => 0 == DateTimeOffset.Compare(x.CreationDate.Date, DateTimeOffset.UtcNow));
+
+
+                if (!request.HrId.HasValue && v != null)
+                {
+                    return new Response()
+                    {
+                        Id = null,
+                        Message = "You already had a test today"
+                    };
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:95]: throw appropriate exception

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 8]
```diff
@@ -24,25 +29,227 @@ public class Handler : IRequestHandler<Request, Response>
...
+                }
+
+                if (request.AssignmentEndDate.Date < DateTimeOffset.UtcNow.Date)
+                {
+                    return new Response()
+                    {
+                        Id = null,
+                        Message = "assignmentEndDate can't be in the past"
+                    };
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:104]: throw appropriate exception

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 9]
```diff
@@ -24,25 +29,227 @@ public class Handler : IRequestHandler<Request, Response>
...
+                }
+
                 var test = _mapper.Map<Test>(request);
 
+                var auditions = await _context.Auditions.AsNoTracking().Where(x => x.Level == request.Level && !x.Deleted).ToListAsync().ConfigureAwait(false);
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs]: !x.Deleted move it to fluent configuration

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 10]
```diff
@@ -24,25 +29,227 @@ public class Handler : IRequestHandler<Request, Response>
...
+
+                if (auditions.Count() < 1)
+                {
+                    return new Response()
+                    {
+                        Id = null,
+                        Message = "Not enough auditions"
+                    };
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:117]: throw appropriate exception

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 11]
```diff
@@ -24,25 +29,227 @@ public class Handler : IRequestHandler<Request, Response>
...
+                }
+
+                var topics = await _context.Topics.AsNoTracking().Where(x => x.Level == request.Level && !x.Deleted).ToListAsync().ConfigureAwait(false);
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs]: !x.Deleted move it

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 12]
```diff
@@ -24,25 +29,227 @@ public class Handler : IRequestHandler<Request, Response>
...
+
+                if (topics.Count() < 2)
+                {
+                    return new Response()
+                    {
+                        Id = null,
+                        Message = "Not enough topics"
+                    };
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:128]: throw appropriate exception

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 13]
```diff
@@ -24,25 +29,227 @@ public class Handler : IRequestHandler<Request, Response>
...
                 _context.Tests.Add(test);
-                await _context.SaveChangesAsync(cancelationtoken);
+
+                if (await CreateTestGrammarQuestionsAsync(test, cancelationtoken).ConfigureAwait(false))
+                {
+                    return new Response()
+                    {
+                        Id = null,
+                        Message = "Not enough Grammar Questions"
+                    };
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:144]: throw appropriate exception

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 14]
```diff
@@ -24,25 +29,227 @@ public class Handler : IRequestHandler<Request, Response>
...
+                }
+                if (await CreateTestAuditionQuestionsAsync(test, cancelationtoken).ConfigureAwait(false))
+                {
+                    return new Response()
+                    {
+                        Id = null,
+                        Message = "Not enough Audition Questions"
+                    };
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:152]: throw appropriate exception

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 15]
```diff
@@ -24,25 +29,227 @@ public class Handler : IRequestHandler<Request, Response>
...
+                .ElementAt(_rand.Next(0, auditions.Count() - 1))
+                .Id;
+            }
+
+            private async Task<bool> CreateTestGrammarQuestionsAsync(Test test, CancellationToken cancelationtoken)
+            {
+                var questions = await _context
+                .Questions
+                .AsNoTracking()
+                .Where(x => x.Level == test.Level && x.AuditionId == null && !x.Deleted).ToListAsync().ConfigureAwait(false);
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs]: !x.Deleted move it

**Review by DShcharbakou — CHANGES_REQUESTED** [review]

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs (hunk 1 above)
Comment (andreyrogov8) [reply Elevel/Application/Features/TestCommands/CreateTestCommand.cs:46]: _userManager.Users.Any(x => x.Id == request.UserId)

File: Elevel/Application/Features/TestCommands/AssignTestCommand.cs [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+                .Id;
+            }
+
+            private async Task<bool> CreateTestGrammarQuestionsAsync(Test test, CancellationToken cancelationtoken)
+            {
+                var questions = await _context
+                .Questions
+                .AsNoTracking()
+                .Where(x => x.Level == test.Level && x.AuditionId == null).ToListAsync().ConfigureAwait(false);
+                if (questions.Count() < 12)
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/AssignTestCommand.cs:132]: magic number

File: Elevel/Application/Features/TestCommands/AssignTestCommand.cs [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+            }
+
+            private async Task<bool> CreateTestAuditionQuestionsAsync(Test test, CancellationToken cancelationtoken)
+            {
+                var questions = await _context
+                .Questions
+                .AsNoTracking()
+                .Where(x => x.Level == test.Level && x.AuditionId == test.AuditionId).ToListAsync().ConfigureAwait(false);
+
+                if (questions.Count() < 10)
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/AssignTestCommand.cs:166]: magic number

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 18]
```diff
@@ -1,48 +1,201 @@
...
             public async Task<Response> Handle(Request request, CancellationToken cancelationtoken)
             {
+                if (!_userManager.Users.Any(x => x.Id == request.UserId))
+                {
+                    throw new NotFoundException("User");
+                }
+
+
+
+
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:51]: a lot of blank lines. It looks not so good.

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 19]
```diff
@@ -1,48 +1,201 @@
...
+            public Handler(IApplicationDbContext context, IMapper mapper, UserManager<ApplicationUser> userManager)
             {
                 _context = context;
                 _mapper = mapper;
+                _userManager = userManager;
+
             }
             public async Task<Response> Handle(Request request, CancellationToken cancelationtoken)
             {
+                if (!_userManager.Users.Any(x => x.Id == request.UserId))
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:44]: use async version

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 20]
```diff
@@ -1,48 +1,201 @@
...
+                {
+                    throw new NotFoundException("User");
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:46]: You need to use message with more details in exception.

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 21]
```diff
@@ -1,48 +1,201 @@
...
+                }
+
+
+
+
+                if (_context.Tests.Any(x => x.UserId == request.UserId
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:52]: use async version

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 22]
```diff
@@ -1,48 +1,201 @@
...
+                {
+                    throw new ValidationException("Not enough Audition Questions");
+                }
+
+                await _context.SaveChangesAsync(cancelationtoken).ConfigureAwait(false);
+
+                var response = _mapper.Map<Response>(test);
+
+                return response;
+
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:95]: blank line does not needed

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 23]
```diff
@@ -1,48 +1,201 @@
...
+                return false;
+            }
+            private async Task<bool> CreateTestAuditionQuestionsAsync(Test test, CancellationToken cancelationtoken)
+            {
+                var questions = await _context
+                .Questions
+                .AsNoTracking()
+                .Where(x => x.Level == test.Level && x.AuditionId == test.AuditionId).ToListAsync().ConfigureAwait(false);
+
+                if (questions.Count() < 10)
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:154]: magic number

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 24]
```diff
@@ -1,48 +1,201 @@
...
+                .ElementAt(_rand.Next(0, auditions.Count() - 1))
+                .Id;
+            }
+            private async Task<bool> CreateTestGrammarQuestionsAsync(Test test, CancellationToken cancelationtoken)
+            {
+                var questions = await _context
+                .Questions
+                .AsNoTracking()
+                .Where(x => x.Level == test.Level && x.AuditionId == null).ToListAsync().ConfigureAwait(false);
+                if (questions.Count() < 12)
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:121]: magic number

File: Elevel/Application/Features/TestCommands/GetAllTestsQuery.cs [hunk 25]
```diff
@@ -39,46 +36,45 @@ public async Task<Response> Handle(Request request, CancellationToken cancelatio
...
+                }
+
 
                 return new Response()
                 {
                     CurrentPage = request.CurrentPage,
                     PageSize = request.PageSize,
                     RowCount = await tests.CountAsync(),
-                    Results = await tests.Skip(request.SkipCount()).Take(request.PageSize).ProjectTo<TestDTO>(_mapper.ConfigurationProvider).ToListAsync()
+                    Results = await tests.Skip(request.SkipCount()).Take(request.PageSize).ProjectTo<TestDTO>(_mapper.ConfigurationProvider).OrderBy(x => x.TestNumber).ToListAsync().ConfigureAwait(false)
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/GetAllTestsQuery.cs:59]: it's not readable. You can use few lines for this expression

File: Elevel/Application/Infrastructure/ValidationException.cs [hunk 26]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace Elevel.Application.Infrastructure
+{
+    class ValidationException: Exception
+    {
+        public ValidationException(string message) 
+            :base(message) 
+        {
+
+        }
```
Comment (DShcharbakou) [inline Elevel/Application/Infrastructure/ValidationException.cs:15]: Best practice is use three overloads
https://docs.microsoft.com/en-us/dotnet/standard/exceptions/how-to-create-user-defined-exceptions

**Review by DShcharbakou — CHANGES_REQUESTED** [review]
One important thing. On real project you shouldn't update packages after each pull request because sometimes not all of packagies will be works fine on new version as worked on a previous version. In this case usually make separete task for updating packagies.

**Review by DShcharbakou — APPROVED** [review]

**Review by 4500zenja1 — APPROVED** [review]

**Pull request merged**